### PR TITLE
Build: allow offline build using system libraries

### PR DESCRIPTION
**Pull request by drewrisinger**
Goal of this PR is to support offline build using a system like https://nixos.org/, which does not support buildtime network access. This prevents the current system of fetching a specific version of pybind11 via git.

Essentially, this refactors the pybind11 library search process (via cmake) to a module, and then sets that module to default to the system pybind11 library. This will only change your build process if you also have pybind11 libraries installed on the path.

I'm also open to a flag to switch behavior between system library vs specified pybind11 version.

### CODE REVIEW

File: pybind_interface/GetPybind11.cmake [hunk 1]
```diff
@@ -0,0 +1,13 @@
+include(FetchContent)
+
+FetchContent_Declare(
+  pybind11
+  GIT_REPOSITORY https://github.com/pybind/pybind11
+  GIT_TAG v2.2.4
+)
+FetchContent_GetProperties(pybind11)
+find_package(pybind11 CONFIG)
+if((NOT pybind11_FOUND) AND (NOT pybind11_POPULATED)) # check first on system path, then attempt git fetch
```
Comment (95-martin-orion) [inline pybind_interface/GetPybind11.cmake:10]: If the local pybind version is older than the requested one, will this recognize that and attempt git-fetching?

**Review by 95-martin-orion — COMMENTED** [review]
This seems reasonable to me, and I support the "try offline first, then fallback to install process" philosophy. Just a couple of notes.

File: pybind_interface/GetPybind11.cmake (hunk 1 above)
Comment (drewrisinger) [reply pybind_interface/GetPybind11.cmake:10]: Fair point. The fix is changing ``find_package(...) -> find_package(pybind11 2.4.4 CONFIG)``, which will do version checking against the installed pybind11 version. For example, here's the section of ``pybind11ConfigVersion.cmake`` checking compatibility of my system ``pybind11 2.6.2`` against a given version (specified as ``PACKAGE_FIND_VERSION`` when doing ``find_package()``).

```cmake
set(PACKAGE_VERSION "2.6.2")

if (PACKAGE_FIND_VERSION_RANGE)
  # Package version must be in the requested version range
  if ((PACKAGE_FIND_VERSION_RANGE_MIN STREQUAL "INCLUDE" AND PACKAGE_VERSION VERSION_LESS PACKAGE_FIND_VERSION_MIN)
      OR ((PACKAGE_FIND_VERSION_RANGE_MAX STREQUAL "INCLUDE" AND PACKAGE_VERSION VERSION_GREATER PACKAGE_FIND_VERSION_MAX)
        OR (PACKAGE_FIND_VERSION_RANGE_MAX STREQUAL "EXCLUDE" AND PACKAGE_VERSION VERSION_GREATER_EQUAL PACKAGE_FIND_VERSION_MAX)))
    set(PACKAGE_VERSION_COMPATIBLE FALSE)
  else()
    set(PACKAGE_VERSION_COMPATIBLE TRUE)
  endif()
else()
  if(PACKAGE_VERSION VERSION_LESS PACKAGE_FIND_VERSION)
    set(PACKAGE_VERSION_COMPATIBLE FALSE)
  else()
    set(PACKAGE_VERSION_COMPATIBLE TRUE)
    if(PACKAGE_FIND_VERSION STREQUAL PACKAGE_VERSION)
      set(PACKAGE_VERSION_EXACT TRUE)
    endif()
  endif()
endif()
```
I'll push this fix.

File: requirements-dev.txt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# test requirements
```
[the remark this answers is not in the document]
Comment (drewrisinger) [reply requirements-dev.txt:1]: I wasn't aware of that. I'm OK dropping the commit if that makes things easier for this PR.

[the remark this answers is not in the document]
Comment (95-martin-orion) [reply requirements-dev.txt:1]: Sounds good to me. (This may also resolve the standing build failures)

**Review by 95-martin-orion — APPROVED** [review]
Thanks for this fix!

**Pull request merged**